### PR TITLE
feat: configure assets in asset canister

### PR DESCRIPTION
**Pull request by smallstepman**
# Description

This PR enables configuring custom HTTP headers and Cache headers for assets in asset canister, using `.ic-assets.json` config file.

Updates `dfx` with:
- the latest `agent-rs` changes:
  - https://github.com/dfinity/agent-rs/pull/352
  - https://github.com/dfinity/agent-rs/pull/358
  - https://github.com/dfinity/agent-rs/pull/365
  - https://github.com/dfinity/agent-rs/pull/362
  - https://github.com/dfinity/agent-rs/pull/361
  - https://github.com/dfinity/agent-rs/pull/368

- and latest `certified-assets` changes:
  - https://github.com/dfinity/certified-assets/pull/31 ([0.2.2 release](https://github.com/dfinity/certified-assets/releases/tag/0.2.2 )) + `ic-wasm --output assetstorage.wasm assetstorage.wasm metadata --file assetstorage.did --visibility public candid:service`
  - https://github.com/dfinity/certified-assets/pull/35 ([0.2.3 release](https://github.com/dfinity/certified-assets/releases/tag/0.2.3)) + `ic-wasm --output assetstorage.wasm assetstorage.wasm metadata --file assetstorage.did --visibility public candid:service`
  - https://github.com/dfinity/certified-assets/pull/36 ([0.2.4 release](https://github.com/dfinity/certified-assets/releases/tag/0.2.4))

- replaces `ic-types` with `candid` 

Closes https://dfinity.atlassian.net/browse/SDK-479.
Closes https://dfinity.atlassian.net/browse/SDK-473.

# How Has This Been Tested?

- created a new dfx project with assets, created `.ic-assets.json` file, deployed, fetched asset in the browser, and verified that the HTTP headers defined in config file are attached to HTTP response for the asset file
- e2e tests

# Checklist:

- [x] run `cargo build` to update `Cargo.lock` (blocked by https://github.com/dfinity/agent-rs/pull/358)
- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation. - please suggest a good place for that

### CODE REVIEW

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

**Review by ericswanson-dfinity — DISMISSED** [review]
Please describe the .ic-asset.json file / functionality in the changelog.  It should be okay to copy the text from the agent-rs changelog.

Also, please add e2e tests that verify / demonstrate the functionality, at least at a basic level.  This should include a couple assets with different cache settings, as well as a project with a frontend that has both `src/<project>_assets/assets/.ic-asset.json` and `src/<project>_assets/src/.ic-asset.json` that each set properties for files in the separate directories, but do not affect each other.

**Comment by smallstepman** [comment]
@ericswanson-dfinity regarding e2e test: how would you advise testing the asset's headers? Should I construct some node script with `agent-js` to query the response headers for assets? e.g.
```
assert_command node get-asset-headers.js http://127.0.0.1:8000/?canisterId=ryjl3-tyaaa-aaaaa-aaaba-cai/logo.png
assert_match 'X-Allow-Origin: "*"' 
``` 
---
Also, need help with fixing this:
- source:
   - e2e: https://github.com/dfinity/sdk/blob/master/e2e/tests-dfx/deploy.bash#L47
   - code: https://github.com/dfinity/agent-rs/blob/6c8686fc5bd410451c2b660f31f947580208e792/ic-asset/src/sync.rs#L65
- error - https://github.com/dfinity/sdk/runs/7237209576?check_suite_focus=true

**Comment by ericswanson-dfinity** [comment]
> @ericswanson-dfinity regarding e2e test: how would you advise testing the asset's headers? 

Try `curl -v`.  There are other tests that use curl.

**Comment by ericswanson-dfinity** [comment]
> Also, need help with fixing this:

The canister is called `hello_frontend` now, not `hello_assets`

**Comment by smallstepman** [comment]
> The canister is called hello_frontend now, not hello_assets
I think the e2e error is unrelated to canister name 

> Try curl -v. There are other tests that use curl.
I couldn't find any tests using curl: https://github.com/dfinity/sdk/search?q=curl
Also, I don't think it's currently possible to fetch the asset from the asset canister directly via HTTP (i.e. without `agent-js`) - tried it some time ago and didn't work, but maybe I was doing something wrong

**Comment by smallstepman** [comment]
wow, it is the worst

**Comment by smallstepman** [comment]
ah, I see, seems like the filename needs to go first in the url for this to work

**Review by sesi200 — APPROVED** [review]

File: src/dfx/Cargo.toml [hunk 1]
```diff
@@ -44,7 +44,7 @@ garcon = { version = "0.2", features = ["async"] }
 hex = {version = "0.4.2", features = ["serde"] }
 humantime = "2.1.0"
 humantime-serde = "1.1.1"
-ic-types = "0.3.0"
+ic-types = "0.4.0"
```
Comment (lwshang) [inline src/dfx/Cargo.toml:47]: Could you check if we can remove direct dependency of `ic-types` here?
Since candid re-export `Pincipal` from `ic-types`

Comment (smallstepman) [reply src/dfx/Cargo.toml]: yup, works :) 
the only thing I'm concerned about is the git diff size for this PR: this change encompasses +50 files https://github.com/dfinity/sdk/pull/2320/commits/fd8344e4fbf3a3c5700a6d8daf33ce40d96b024e

@ericswanson-dfinity let me know if you'd want to see this reverted and moved to separate PR

File: e2e/tests-dfx/assetscanister.bash [hunk 2]
```diff
@@ -333,3 +333,161 @@ CHERRIES" "$stdout"
...
+    export CURL_ME="http://localhost:$PORT/index.html?canisterId=$ID"
+    echo $CURL_ME
+    assert_command curl --head "$CURL_ME"
+    # shellcheck disable=SC2154
+    assert_match "x-header: x-value"
+
+    # assert_not_match '"/will-delete-this.txt"'
+}
+
+@test "another anotanotherher asset configuration via .ic-assets.json" {
```
Comment (ericswanson-dfinity) [inline e2e/tests-dfx/assetscanister.bash:464]: Would you please update these test names to something that reflects what they are testing?  And remove the `echo` statements and whatever else isn't necessary for the test.

File: CHANGELOG.adoc [hunk 3]
```diff
@@ -3,6 +3,8 @@
 
 = UNRELEASED
 
+=== Updated agent-rs to 0.19.0
```
Comment (ericswanson-dfinity) [inline CHANGELOG.adoc:6]: Please follow the pattern used by previous versions, where changes like this go in a `Dependencies` section.

File: e2e/tests-dfx/assetscanister.bash (hunk 2 above)
Comment (smallstepman) [reply e2e/tests-dfx/assetscanister.bash:464]: sorry, this is a work in progress, I couldn't make this pass locally so I'm just spamming GH Actions with a battery of tests  with slight changes

File: CHANGELOG.adoc (hunk 3 above)
Comment (smallstepman) [reply CHANGELOG.adoc]: forgot about adding some info about `ignore` field

File: CHANGELOG.adoc [hunk 4]
```diff
@@ -3,8 +3,53 @@
...
+        "match": "**/*",
+        "headers": null
+    },
+    {
+        "match": "file.json",
+        "ignore": true
+    }
+]
+----
+- Configuring assets works only during asset creation.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.adoc:38]: Let's call out a a little more visibly that at present, changes to these fields don't take effect once the asset is created, and that we will follow up with improvements to handle updating these properties.

File: CHANGELOG.adoc [hunk 5]
```diff
@@ -3,8 +3,53 @@
...
+- `headers` from multiple applicable rules are being stacked/concatenated, unless `null` is specified, which resets/empties the headers.
+- Both `"headers": {}` and absence of `headers` field don't have any effect on end result.
+- Valid JSON format is required, i.e. the array of maps, `+match+` field is required. Only the following fields are accepted: `+cache+`, `+ignore+`, `+headers+`, `+match+`. The glob pattern has to be valid.
+- The way matching rules work:
++
+. The most deeply nested config file takes precedens over the one in parent dir. In other words, properties from a rule matching a file in a subdirectory override properties from a rule matching a file in a parent directory
+. Order of rules withing file matters - last rule in config file takes precedens over the first one
```
Comment (ericswanson-dfinity) [inline CHANGELOG.adoc:45]:
```suggestion
. Order of rules within file matters - last rule in config file takes precedence over the first one
```

File: CHANGELOG.adoc [hunk 6]
```diff
@@ -3,8 +3,53 @@
...
++
+- The way `+ignore+` field works:
++
+. If a directory is ignored, file and directories within it cannot be un-ignored.
+. A file can be ignored and un-ignored many times, as long as it's parent directory hasn't been ignored.
```
Comment (ericswanson-dfinity) [inline CHANGELOG.adoc:50]:
```suggestion
. A file can be ignored and un-ignored many times, as long as any of its parent directories haven't been ignored.
```

File: CHANGELOG.adoc [hunk 7]
```diff
@@ -3,8 +3,53 @@
...
+- `headers` from multiple applicable rules are being stacked/concatenated, unless `null` is specified, which resets/empties the headers.
+- Both `"headers": {}` and absence of `headers` field don't have any effect on end result.
+- Valid JSON format is required, i.e. the array of maps, `+match+` field is required. Only the following fields are accepted: `+cache+`, `+ignore+`, `+headers+`, `+match+`. The glob pattern has to be valid.
+- The way matching rules work:
++
+. The most deeply nested config file takes precedens over the one in parent dir. In other words, properties from a rule matching a file in a subdirectory override properties from a rule matching a file in a parent directory
+. Order of rules withing file matters - last rule in config file takes precedens over the first one
++
+- The way `+ignore+` field works:
++
```
Comment (ericswanson-dfinity) [inline CHANGELOG.adoc:48]: Let's call out this behavior, too:
```suggestion
- The way `+ignore+` field works:
+
. By default, files that begin with a `.` are ignored, while all other files are included.
. The `.ignore` field overrides this, if present.
```

File: e2e/assets/assetscanister/src/e2e_project_frontend/assets/.ic-assets.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    "match": "*",
+    "cache": {
+      "max_age": 500
+    },
+    "headers": {
+      "x-header": "x-value"
+    }
+  },
+  {
+    "match": ".*",
```
Comment (ericswanson-dfinity) [inline e2e/assets/assetscanister/src/e2e_project_frontend/assets/.ic-assets.json:16]: Can we target this at a specific file or directory?  One interpretation of ".*" as a regex is "match any character, 0 or more times."  But here I think we mean "match a dot, followed by anything or nothing."

So I'm also I'm wondering, and would like to see the test demonstrate: which is the behavior of the glob library (and therefore dfx)?

File: e2e/tests-dfx/assetscanister.bash (hunk 2 above)
Comment (ericswanson-dfinity) [reply e2e/tests-dfx/assetscanister.bash:464]: Understood.  No problem!

File: e2e/assets/assetscanister/src/e2e_project_frontend/assets/.ic-assets.json (hunk 8 above)
Comment (smallstepman) [reply e2e/assets/assetscanister/src/e2e_project_frontend/assets/.ic-assets.json:16]: > But here I think we mean "match a dot, followed by anything or nothing."

correct

We're currently limited until this issue is resolved https://github.com/BurntSushi/ripgrep/issues/2229
Once resolved, the user will be able to use `.` as regex when `ignore` field won't be set, and as a "match a dot" if `ignore: false`

File: CHANGELOG.adoc [hunk 9]
```diff
@@ -3,8 +3,55 @@
...
 
+== Dependencies
+
+=== Updated `+agent-rs+` to 0.20.0
+=== Updated `+candid+` to 0.7.15
+
 == DFX
 
+=== feat: support for configuring assets in assets canister
+- The `+.ic-assets.json+` file should be placed inside directory with assets, or it's subdirectories. Multiple config files can be used (nested in subdirectories). Example of `.ic-assets.json` file format:
```
Comment (ericswanson-dfinity) [inline CHANGELOG.adoc:14]:
```suggestion
- The `+.ic-assets.json+` file should be placed inside directory with assets, or its subdirectories. Multiple config files can be used (nested in subdirectories). Example of `.ic-assets.json` file format:
```

**Pull request merged**